app/Actions/Photo/Archive.php

```diff
@@ -28,74 +60,82 @@ public function __construct()
+	 *                         {@link Archive::SMALL2X},
+	 *                         {@link Archive::SMALL},
+	 *                         {@link Archive::THUMB2X},
+	 *                         {@link Archive::THUMB}
+	 *
+	 * @return Response
 	 */
-	public function do(array $photoIDs, $kind_request)
+	public function do(array $photoIDs, string $variant): Response
 	{
 		if (count($photoIDs) === 1) {
-			$response = $this->file($photoIDs[0], $kind_request);
+			$response = $this->file($photoIDs[0], $variant);
 		} else {
-			$response = $this->zip($photoIDs, $kind_request);
+			$response = $this->zip($photoIDs, $variant);
 		}
 
 		return $response;
 	}
 
-	public function file($photoID, $kind_request)
+	protected function file($photoID, $variant): BinaryFileResponse
 	{
-		$ret = $this->extract_names($photoID, $kind_request);
-		if ($ret === null) {
-			return abort(404);
+		$archiveFileInfo = $this->extractFileInfo($photoID, $variant);
+		if ($archiveFileInfo === null) {
+			abort(404);
 		}
+		$response = new BinaryFileResponse($archiveFileInfo->getFullPath());
 
-		list($title, $kind, $extension, $url) = $ret;
-
-		// Set title for photo
-		$file = $title . $kind . $extension;
-
-		$response = new BinaryFileResponse($url);
-
-		return $response->setContentDisposition(ResponseHeaderBag::DISPOSITION_ATTACHMENT, $file);
+		return $response->setContentDisposition(
+			ResponseHeaderBag::DISPOSITION_ATTACHMENT,
+			$archiveFileInfo->getFilename()
+		);
 	}
 
-	public function zip(array $photoIDs, string $kind_request)
+	protected function zip(array $photoIDs, string $variant): StreamedResponse
 	{
-		$response = new StreamedResponse(function () use ($kind_request, $photoIDs) {
+		$response = new StreamedResponse(function () use ($variant, $photoIDs) {
 			$options = new \ZipStream\Option\Archive();
 			$options->setEnableZip64(Configs::get_value('zip64', '1') === '1');
 			$zip = new ZipStream(null, $options);
 
-			$files = [];
+			$fileCounter = [];
 			foreach ($photoIDs as $photoID) {
-				$ret = $this->extract_names($photoID, $kind_request);
-				if ($ret == null) {
-					return abort(404);
+				$archiveFileInfo = $this->extractFileInfo($photoID, $variant);
+				if ($archiveFileInfo == null) {
+					abort(404);
 				}
 
-				list($title, $kind, $extension, $url) = $ret;
-
 				// Set title for photo
-				$file = $title . $kind . $extension;
+				$filename = $archiveFileInfo->getFilename();
 				// Check for duplicates
-				if (!empty($files)) {
-					$i = 1;
-					$tmp_file = $file;
-					while (in_array($tmp_file, $files)) {
-						// Set new title for photo
-						$tmp_file = $title . $kind . '-' . $i . $extension;
-						$i++;
-					}
-					$file = $tmp_file;
+				if (array_key_exists($filename, $fileCounter)) {
+					$cnt = $fileCounter[$filename];
+					$cnt++;
+					$fileCounter[$filename] = $cnt;
+					$filename = $archiveFileInfo->getFilename('-' . $cnt);
+				} else {
+					$fileCounter[$filename] = 1;
```

Comment:
That's true. My code generates `original.jpg` followed by `ogirinal-2` if the same file name appears a second time. But your old code would have done the same (at least for the very first time), because `empty($files)` would have been true, too.

Anyway, we should decide what we want.

 1. Leave it like that. I think it is not so bad and very much what must file managers also do, if you copy/move a file to destination which already contains a file with the same name.
 2. Change the code such that `-1` is always added to the original file name, even if there is no duplicate. I don't know if this is desirable.
 3. Only add `-1` to a filename if there are also `-2`, `-3` and so on, but don't add anything if the filename is unique. Surely, this would be the most pleasant solution, but it requires to process the list of files two times. At the moment, the code iterates over the list of files only once. Hence, if a filename is seen for the very first time, the code cannot know, if there will be the same filename further down the list. Hence, we cannot decide, if we must add `-1` to it or not. If wanted to decide that, we had to scan the list of files one time, before we actually start to zip the files. I don't know if this additional complexity is worth it.

Reply:
Given your last example in which picture1-2 actually exists as an independent photo in the album I have to revise my code again. In this case the generated `picture1-2` would overwrite the original `picture1-2` which is obviously undesired.

Actually, we should have a test for that. :see_no_evil:

Reply:
I solved the problem. I hope it should be fine now and still a little bit more efficient than the previous code. And much more documented. :sunglasses: